SwiftCompilerSources/Sources/SIL/ApplySite.swift

```diff
@@ -12,60 +12,81 @@
 
 import SILBridging
 
-public struct ApplyOperands {
-  public static let calleeOperandIndex: Int = 0
+/// Argument conventions indexed on an apply's operand.
+///
+/// `partial_apply` operands correspond to a suffix of the callee
+/// arguments.
+///
+/// Example:
+/// ```
+/// func callee(a, b, c, d, e) { }
+///
+/// %pa = partial_apply @callee(c, d, e)
+/// // operand indices:         1, 2, 3
+/// // callee indices:          2, 3, 4
+///
+///  %a = apply         %pa    (a, b)
+/// // operand indices:         1, 2
+/// // callee indices:          0, 1
+/// ```
+public struct ApplyOperandConventions : Collection {
+  public static let calleeIndex: Int = 0
   public static let firstArgumentIndex = 1
+
+  /// Callee's argument conventions indexed on the function's arguments.
+  public let calleeArgumentConventions: ArgumentConventions
+
+  public let unappliedArgumentCount: Int
+
+  public var appliedArgumentCount: Int {
+    calleeArgumentConventions.count - unappliedArgumentCount
+  }
+
+  public func isCallee(operand: Operand) -> Bool {
+    return operand.index == ApplyOperandConventions.calleeIndex
+  }
+
+  public var startIndex: Int { ApplyOperandConventions.firstArgumentIndex }
+
+  public var endIndex: Int { ApplyOperandConventions.firstArgumentIndex + appliedArgumentCount }
+
+  public func index(after index: Int) -> Int {
+    return index + 1
+  }
+
+  public subscript(_ index: Int) -> ArgumentConvention {
```

Comment:
can you rename `index` -> `operandIndex`, so that it's clear that it is not the argument index?

Reply:
Done. There is still no subscript label, because it can only by used to subscript a collection of operand conventions: 
`operandConventions[operand.index]`